src/WinRT.Interop.Generator/Helpers/GuidGenerator.cs

```diff
@@ -0,0 +1,342 @@
+                if (typeDefinition.IsClass)
+                {
+                    return typeDefinition.IsDelegate
+                        ? "delegate({" + GetGuid(typeSignature, interopReferences) + "})" // Delegate case
+                        : GetDefaultInterfaceSignatureFromAttribute(typeDefinition, interopReferences, out TypeSignature defaultInterfaceSig)
+                            ? "rc(" + typeFullName + ";" + GetSignature(defaultInterfaceSig, interopReferences, useWindowsUIXamlProjections) + ")" // Class case with default interface
+                            : "{" + GetGuid(typeSignature, interopReferences) + "}"; // Class case without default interface
+                }
+                if (typeDefinition.IsInterface) // interface case
+                {
+                    return "{" + GetGuid(typeSignature, interopReferences) + "}";
+                }
+
+                throw new ArgumentException("Invalid ElementType.Class");
+
+            case ElementType.SzArray:
+                SzArrayTypeSignature arrayTypeSignature = (SzArrayTypeSignature)typeSignature;
+
+                if (arrayTypeSignature != null)
+                {
+                    return "pinterface({61c17707-2d65-11e0-9ae8-d48564015472};" + GetSignature(arrayTypeSignature.BaseType, interopReferences, useWindowsUIXamlProjections) + ")";
+                }
+
+                throw new ArgumentException("Invalid ElementType.SzArray");
+
+            default:
+                // TODO: throw new ArgumentException("Unsupported ElementType: " + typeSignature.ElementType + " : " + typeSignature.FullName);
+                return typeSignature.FullName;
+        }
+#pragma warning restore IDE0010 // Add missing cases
+    }
+
+    /// <summary>
+    /// Resolves a GUID for the specified type signature by checking well-known WinRT interfaces
+    /// and, if necessary, the type's <c>System.Runtime.InteropServices.GuidAttribute</c>.
+    /// </summary>
+    /// <param name="typeSig">
+    /// The <see cref="AsmResolver.DotNet.Signatures.TypeSignature"/> to resolve to a GUID.
+    /// </param>
+    /// <param name="interopReferences">
+    /// The <see cref="InteropReferences"/> instance to use.
+    /// </param>
+    /// <returns>
+    /// The resolved <see cref="Guid"/> if found.
+    /// </returns>
+    /// <exception cref="ArgumentException">Thrown when the type has no GUID.</exception>
+    private static Guid GetGuid(TypeSignature typeSig, InteropReferences interopReferences)
+    {
+        ITypeDefOrRef typeDefOrRef = typeSig switch
+        {
+            SzArrayTypeSignature szArrayTypeSignature => szArrayTypeSignature.GetUnderlyingType().ToTypeDefOrRef(),
+            GenericInstanceTypeSignature genericSignature => genericSignature.GenericType,
+            _ => typeSig.ToTypeDefOrRef(),
+        };
+        if (WellKnownInterfaceIIDs.try_GetGUID(typeDefOrRef, interopReferences, out Guid result))
+        {
+            return result;
+        }
```

Comment:
Not sure I see how this is correct. Let's say you call this with the type signature being `EventHandler<string>`. Here you're unwrapping it and just calling `try_GetGUID` with the generic type definition. So that will succeed and give you the PIID for `EventHandler<T>`. And this method will return that. But that will **not** be the IID for the (constructed) `EventHandler<string>` signature that we actually have here, no?